src/platforms/evdev-rs/src/lib.rs

```diff
@@ -1,25 +1,1443 @@
+    #[derive(Debug, Copy, Clone, PartialEq, Eq)]
+    pub enum MirPointerAcceleration {
+        mir_pointer_acceleration_none = 1,
+        mir_pointer_acceleration_adaptive = 2,
+    }
+
+    #[repr(i32)]
+    #[derive(Debug, Copy, Clone, PartialEq, Eq)]
+    enum MirPointerAction {
+        mir_pointer_action_button_up,
+        mir_pointer_action_button_down,
+        mir_pointer_action_enter,
+        mir_pointer_action_leave,
+        mir_pointer_action_motion,
+        mir_pointer_actions,
+    }
+
+    #[repr(i32)]
+    #[derive(Debug, Copy, Clone, PartialEq, Eq)]
+    enum MirPointerAxisSource {
+        mir_pointer_axis_source_none,
+        mir_pointer_axis_source_wheel,
+        mir_pointer_axis_source_finger,
+        mir_pointer_axis_source_continuous,
+        mir_pointer_axis_source_wheel_tilt,
+    }
+
+    #[repr(u32)]
+    #[derive(Debug, Copy, Clone, PartialEq, Eq)]
+    enum MirPointerButton {
+        mir_pointer_button_primary = 1,
+        mir_pointer_button_secondary = 2,
+        mir_pointer_button_tertiary = 4,
+        mir_pointer_button_back = 8,
+        mir_pointer_button_forward = 16,
+        mir_pointer_button_side = 32,
+        mir_pointer_button_extra = 64,
+        mir_pointer_button_task = 128,
+    }
+
+    #[repr(i32)]
+    #[derive(Debug, Copy, Clone, PartialEq, Eq)]
+    enum MirPointerHandedness {
+        mir_pointer_handedness_right = 0,
+        mir_pointer_handedness_left = 1,
+    }
+
+    #[derive(Copy, Clone)]
+    struct PointerSettingsC {
+        pub handedness: i32,
+        pub cursor_acceleration_bias: f64,
+        pub acceleration: i32,
+        pub horizontal_scroll_scale: f64,
+        pub vertical_scroll_scale: f64,
+    }
+
+    #[derive(Copy, Clone)]
+    struct PointerSettingsRs {
+        pub is_set: bool,
+        pub handedness: i32,
+        pub cursor_acceleration_bias: f64,
+        pub acceleration: i32,
+        pub horizontal_scroll_scale: f64,
+        pub vertical_scroll_scale: f64,
+        pub has_error: bool
+    }
+
+    extern "Rust" {
+        type PlatformRs;
+        type DeviceObserverRs;
+        type InputDeviceRs<'fd>;
+        type InputDeviceInfoRs;
+
+        fn start(self: &mut PlatformRs);
+        fn continue_after_config(self: &PlatformRs);
+        fn pause_for_config(self: &PlatformRs);
+        fn stop(self: &mut PlatformRs);
+        fn create_device_observer(self: &PlatformRs) -> Box<DeviceObserverRs>;
+        fn create_input_device(self: &PlatformRs, device_id: i32) -> Box<InputDeviceRs<'_>>;
+
+        fn activated(self: &mut DeviceObserverRs, fd: i32);
+        fn suspended(self: &mut DeviceObserverRs);
+        fn removed(self: &mut DeviceObserverRs);
+
+        /// # Safety
+        ///
+        /// This is unsafe because it is receiving raw pointers from C++ as parameters.
+        unsafe fn start(
+            self: &mut InputDeviceRs,
+            input_sink: *mut InputSink,
+            event_builder: *mut EventBuilder,
+        );
```

Comment:
Why isn't this

```suggestion
        fn start(
            self: &mut InputDeviceRs,
            input_sink: &mut InputSink,
            event_builder: &mut EventBuilder,
        );
```
It's not clear to me why we're using raw pointers (and hence `unsafe`) here, rather than references?

Reply:
This is a direct translate of the C++ code, which uses raw pointers.